charon-ml/src/Substitute.ml

```diff
@@ -69,15 +101,37 @@ let subst_at_binder_zero (subst : single_binder_subst) : subst =
     tr_self = subst.tr_sb_self;
   }
 
+(** Substitute the variables bound by the current (level 0) binder, and shift variables to remove the current binder level. *)
+let subst_remove_binder_zero (subst : single_binder_subst) : subst =
+  let subst_remove_zero subst nosubst = function
+    | Bound (0, id) -> subst id
+    | Bound (dbid, varid) when dbid > 0 -> nosubst (Bound (dbid - 1, varid))
```

Comment:
Nit: you actually don't need the `when dbid > 0`

Reply:
Yes I do, if we're already inside a binder when we start substituting, variables pointing to binders in-scope when we start will get negative indices when we dig inside other binders. Only the variables with positive indices correspond to binders deeper than where we started and thus that need to be shifted.

Reply:
Oh indeed, I forgot about the negative binders. That would probably be worth a comment in the future.